### ISSUE TITLE
Persistent SSH session

### ISSUE THREAD

**Issue by davidanthoff** [microsoft/vscode-remote-release#3096]
A lot of users of our [Julia extension](https://marketplace.visualstudio.com/items?itemName=julialang.language-julia) love the remote SSH features, they are super useful in a scientific research computing context!

One thing that has come up a lot lately, though, is that folks want an ability to have a persistent session on a server that they can disconnect from and then reconnect to way later.

The scenario is essentially something like this:
1. User connects to some remote server via the Remote SSH extension
2. They do some work
3. They get interrupted and close their local VS Code instance
4. They do something else for a couple of days
5. The now want to reconnect to the server, and ideally _nothing_ was shut down on the server. For example, if they opened a REPL in step 2 and started a Julia instance in that REPL, it would still be running, and was never shut down. Similarly with pretty much all other runtime state they had.

I think this might be related to https://github.com/microsoft/vscode-remote-release/issues/440, but what I'm really asking for is not just an option to change the timeout, but some way to make the timeout infinit, so that the server session never shuts down.

This might also require some kind of UI, maybe a command like "Disconnect but keep server alive" or something like that, not sure.

**Comment by xgdgsc** [comment]
https://github.com/microsoft/vscode-remote-release/issues/1790 also related.

**Comment by EricChen1248** [comment]
Not a perfectly solution, but what about opening the terminal into [Screen](https://linux.die.net/man/1/screen). Users could reconnected back to their screen session each time, also gives more fine grain control over resource releases (properly terminating a process vs an always persistent session)

**Comment by PavelSosin-320** [comment]
This is a well-known pattern in the Cloud Applications. The problem is always that lifecycles of Server and client are not synchronized and either server or client or connection can go down at any point all clients will hang. Then, clients have to re-authenticate. If the client is idle at the point of time was idle its user will have no knowledge of how to proceed. To avoid such a died-end situation client needs awareness of the "session status"  i.e. Server needs session management and ability to report either current session status to the client, at least, by the explicit client's request or every re-connect request coming from the client has had session attribute. There is also a security risk that the session can be hijacked by anybody who has a valid login of the server when the client was idle.

**Comment by janosh** [comment]
Glad I found this! Just wanted to mention that I believe this is far larger in scope than the Julia extension. I bet most people forced to work from home these days would really appreciate this feature!

**Comment by bobaoapae** [comment]
This is what i need for stop using RDP to remote developmet. Sometimes i need keep service runing for several days and i don't want keep my own pc on just for it. Keep process running without vscode remain connected and re-establish connection/state after re-open vscode will be amazing!

**Comment by tom-programming** [comment]
This would be a real game changer for myself, and I believe, a whole lot of people. Now I use tmux to achieve the same but it is uncomfortable in many ways

**Comment by mineralres** [comment]
Glad to hear that this feature request received, it would be very useful for my working context.

**Comment by ImMrMa** [comment]
Is there any progress on this feature recently?

**Comment by abhijithda** [comment]
Hopefully the timeout considers people coming back to work from long weekends, and resuming/waking up a sleeping computer/laptop :)

**Comment by taiya** [comment]
> This would be a real game changer for myself, and I believe, a whole lot of people. Now I use tmux to achieve the same but it is uncomfortable in many ways

Agreed, e.g. tmux interferes with mouse scrolling, and does 100x other things on top of having just a persistent SSH session. Considering vscode has a server running remotely... it shouldn't be impossible to have a `keep-server-alive` mode?

**Comment by ImMrMa** [comment]
It has been nearly two years since this issue opened. Could we pay a little attention to solve the problem? I think it is a great improvement to users.

**Comment by lucasresck** [comment]
Is there a workaround for persistent Jupyter Notebooks through SSH connection? After reconnecting to remote, I would like to reconnect to my Jupyter kernel, or at least to keep the notebooks running.

**Comment by fpickard** [comment]
+1 for this feature request.  It would be great to have native ssh session persistence as a first class feature, instead of DIY via screen/tmux as others have commented.

**Comment by march1993** [comment]
+1 for this feature request. I have to re-open several terminals and each different commands in each window, which is quite annoying everytime I launch vscode and open the project.

**Comment by nkutbw** [comment]
> Is there a workaround for persistent Jupyter Notebooks through SSH connection? After reconnecting to remote, I would like to reconnect to my Jupyter kernel, or at least to keep the notebooks running.

You could combine nohup and port mapping. More detail could be seen [here](https://zhuanlan.zhihu.com/p/367020783)

**Comment by nkutbw** [comment]
Now I am considering using nohup and port mapping to open a [code-server](https://github.com/cdr/code-server), which could be seen as a browser vscode. I prefer vscode to jupyter lab mainly for its vim extension. If the **Persistent SSH session** function is implemented locally, you will be the best ide in my heart.👍

**Comment by lucasresck** [comment]
> You could combine nohup and port mapping. More detail could be seen [here](https://zhuanlan.zhihu.com/p/367020783)

Hi there, thanks for your answer. I think this applies to standard Jupyter Notebook and JupyterLab through CLI, right? I was thinking more about a way to have this working for notebooks inside VS Code too. I edited my comment to specify the question for VS Code.

**Comment by lucasresck** [comment]
> Now I am considering using nohup and port mapping to open a [code-server](https://github.com/cdr/code-server), which could be seen as a browser vscode.

In this case, the VS Code server would be opened in the remote machine and you would access this through standard SSH connection? Interesting idea.

Why using nohup instead of screen?

**Comment by nkutbw** [comment]
> > Now I am considering using nohup and port mapping to open a [code-server](https://github.com/cdr/code-server), which could be seen as a browser vscode.
> 
> In this case, the VS Code server would be opened in the remote machine and you would access this through standard SSH connection? Interesting idea.
> 
> Why using nohup instead of screen?
Thank you. And as a non root user, it is more convenient for me to use nohup. Anyway, we all look forward to a Persistent SSH session in vscode internally just like a terminal or notebook opened in a "nohup" jupyter.

**Comment by mcpeixoto** [comment]
@roblourens is there an update on this feature? This would be a game-changer, please implement this!

**Comment by LeanderK** [comment]
I would also really like to see this! I always work remote and the loss of the session when my computer goes into sleep just drives me mad.

**Comment by shengbeiniao** [comment]
Mark

**Comment by lifeatthesharpend** [comment]
I know I'm coming in late to this thread, but here goes. I'm probably a newb so please be patient. This seems to be related to a few other closed issues mainly:
- Retain remote session when coming back to VSCode from computer sleep #3355 
- Allow session detach & re-attach over long periods of time #1701 
- Custom reconnect timeout #1185 
- Long running kernels #3998 
few others that have closed as well. 

My questions are: 
1. is there a workaround or solution in place yet? 
2. from what I can tell some are using tmux. I know how to use tmux from the terminal but not exactly sure how to best set it up using VS-code. Is there a step-by-step instructions for how to use tmux with vs-code? 
3. I believe that @roblourens mentioned a 3 hours reconnect. Is there a setting for this can this be extended to longer, for example 24 hours? 
4. Is there a way to put the vs-code session in background mode then comeback to it? For example start a python debug run, detach, then travel with the computer, then the next day open , restart computer then re-attach? All this without forcefully disconnection? 
5. Are these features in work? 

btw , why all the thumbs down on this post? 
L

**Comment by coconutattitude** [comment]
How it's going with this 2.5 year-old issue?

**Comment by chang803** [comment]
It would be really useful if we could have this configurable timeout reconnect feature. I believe right now it's 8 hours, which is already increased from 3 hours
But I launch long simulation(>10 hours) through vscode-remote before I go to sleep, when I come back they are killed.
I understand tmux can help with the issue but I have to talk to IT and a lot of other folks to get it installed on the remote machine.

I would really appreciate the configurable reconnect time feature.
Thanks!

**Comment by xgdgsc** [comment]
Yes. I don' t understand how is it so hard to change a single integer before a setting is added. Please change to 168.

**Comment by chang803** [comment]
Any update please?

**Comment by shittiidavincii** [comment]
+1 for an update, persistent ssh using screen or the like would be great.

**Comment by Boris94cr** [comment]
Any update on this feature?

**Comment by ferreirafabio** [comment]
Any update on this feature? Thumb me down as much as you want but we need this feature.

**Comment by holland11** [comment]
+1. Currently using `screen` to achieve something similar, but there are a lot of hoops to jump through to get it working. And the most recent vscode update seems to have broken my `screen` process (now, whenever the vscode session times out while I am attached to a `screen` session in the vscode terminal, the `screen` session becomes unresponsive when I try to reconnect to it and eventually it says it's dead; it used to just let me reconnect to it without issue; going to downgrade my vscode to hopefully get it working again).

For anyone looking to try using `screen` within the remote vscode terminal, one thing you should be aware of is that to get the `code <file>` functionality to work (so that you can use the `code` command in the vscode terminal to open remote files), vscode sets up some EVs in its native terminal session. 
```
VSCODE_GIT_IPC_HANDLE
VSCODE_IPC_HOOK_CLI
```
When you first create a `screen` session through the vscode remote terminal, it will inherit those EVs and so you will still be able to use `code` to open files. However, once the vscode session times out and you reconnect to it, this will generate new EVs and invalidate the old ones. Then you reconnect to the old `screen` session with its outdated vscode EVs and `code <file>` will throw an error. To fix this, you need to refresh the `screen` session's EVs (specifically the ones I mentioned above).

I have this happen automatically with a pair of shell functions (zsh, but might work with bash as well).
```
update_screen_evs() {
  if [[ -f ~/screen_logs/temp_ipc_handle ]] && [[ -f ~/screen_logs/temp_ipc_hook ]]; then
    value=$(<~/screen_logs/temp_ipc_handle)
    export VSCODE_GIT_IPC_HANDLE="$value"
    value=$(<~/screen_logs/temp_ipc_hook)
    export VSCODE_IPC_HOOK_CLI="$value"
    echo "EVs imported."
    return
  fi
  echo "EV files not found."
}

scrn() {
  if [[ $# != 1 ]]; then
    echo "Must provide screen name."
    command screen -ls
    if [[ $STY ]]; then
      echo "Current screen session: $STY"
    fi
    return
  fi
  command echo "$VSCODE_GIT_IPC_HANDLE" > ~/screen_logs/temp_ipc_handle
  command echo "$VSCODE_IPC_HOOK_CLI" > ~/screen_logs/temp_ipc_hook
  command screen -S $1 -X stuff 'update_screen_evs
  '
  command screen -RDS $1
}
```

Using `scrn <session_name>` will reconnect to an existing session named `<session_name>` if it exists (or will create it if it doesn't). Before doing that though, it will copy the native terminal's vscode EVs to files located in the `~/screen_logs/` directory (so you'll probably need to manually create this directory first) and send the `update_screen_evs` command to that session (if the session exists). 

(The `update_screen_evs` command is sent as raw text with an explicit RETURN character. So if that screen session already had text in its user input buffer, the text will have `update_screen_evs` concatenated with it before being returned. Which will likely fail. So in these cases, you'll have to manually run `update_screen_evs` upon reconnecting to the session.)

**Comment by apizz** [comment]
How can we get this moved from the backlog into development?

**Comment by x66ccff** [comment]
Any update on this feature?

**Comment by kingpalethe** [comment]
This is the single feature as far as I can tell where cloud IDE solutions like cloud9 have some significant advantage over VSCode. Would love to figure out how to do this with VScode instead.

**Comment by Kaiserouo** [comment]
Uh so as far as I can tell there is no VSCode solution for this yet? I guess screen or tmux could work but if this can work under all those terminal tabs it would be perfect...

**Comment by TatIwamoto** [comment]
Shouldn't require tmux or screen to get the sessions persistent across disconnects.

**Comment by aminfardi-CD** [comment]
+1 for wanting a native solution on this. Can't get tmux to work for Jupyter notebooks in remote sessions.

**Comment by matriim** [comment]
I just went through the trouble of setting everything up just to find out at the last step that there is no way to let my jupyter notebook run remotely when I close my local machine. I don't know if I should cry or laugh. Please give us this possibility

**Comment by ltalirz** [comment]
This is by far the top-voted issue on vscode-remote-release now

<img width="696" alt="image" src="https://github.com/microsoft/vscode-remote-release/assets/1053245/8a9a3c54-9028-4415-a2a8-637a471d389c">

Message from user community seems clear

**Comment by aiojfunbf** [comment]
Hi, is there any solution for it now?

**Comment by cqlc94** [comment]
Any update now? Would be extremely useful

**Comment by Phil-Garmann** [comment]
@AgFlore can you verify that this workaround works? At least from my testing, it does not seem to remove the necessary argument.
One comment: `./vscode-server` should probably be `/.vscode-server`

**Comment by ma-sadeghi** [comment]
> One workround for now (only for users who wants the remote-server persistent FOREVER, not for users who wish to change the timeout value):
> 
> vscode starts the remote server by launching `sh <vscode-server-bin-path>/bin/code-server --start-server --host=127.0.0.1 --accept-server-license-terms --enable-remote-auto-shutdown --port=0 --telemetry-level all --connection-token-file <path-to-token-file>` in your default shell (`/bin/bash` in my case) after it established the SSH connection.
> 
> Therefore, just remove the `--enable-remote-auto-shutdown` argument, and the server will run without the idle timeout. This can be achieved by a customised wrapper for `sh`. My version goes like:
> 
> ```shell
> #! /bin/bash
> 
> args="$@"
> 
> ignore_me='--enable-remote-auto-shutdown'
> 
> if echo "$@" | grep -q -- "/.vscode-server"; then
>     if echo "$@" | grep -q -- "$ignore_me"; then
>         args=$(echo "$@" | sed "s/$ignore_me//g")
>     fi
> fi
> 
> /usr/bin/sh $args
> ```
> 
> And then `export PATH=<path-where-you-put-that-wrapper>:$PATH` in `.bashrc` (or into vscode's ssh-config, if you dislike polluting your shell environment). Now vscode will launch the remote-server without enabling the annoying idle-shutdown.

Neat! Does this work even when VS Code client is closed?

**Comment by HarunKilic** [comment]
> One workround for now (only for users who wants the remote-server persistent FOREVER, not for users who wish to change the timeout value):
> 
> vscode starts the remote server by launching `sh <vscode-server-bin-path>/bin/code-server --start-server --host=127.0.0.1 --accept-server-license-terms --enable-remote-auto-shutdown --port=0 --telemetry-level all --connection-token-file <path-to-token-file>` in your default shell (`/bin/bash` in my case) after it established the SSH connection.
> 
> Therefore, just remove the `--enable-remote-auto-shutdown` argument, and the server will run without the idle timeout. This can be achieved by a customised wrapper for `sh`. My version goes like:
> 
> ```shell
> #! /bin/bash
> 
> args="$@"
> 
> ignore_me='--enable-remote-auto-shutdown'
> 
> if echo "$@" | grep -q -- "/.vscode-server"; then
>     if echo "$@" | grep -q -- "$ignore_me"; then
>         args=$(echo "$@" | sed "s/$ignore_me//g")
>     fi
> fi
> 
> /usr/bin/sh $args
> ```
> 
> And then `export PATH=<path-where-you-put-that-wrapper>:$PATH` in `.bashrc` (or into vscode's ssh-config, if you dislike polluting your shell environment). Now vscode will launch the remote-server without enabling the annoying idle-shutdown.

Could you explain this a bit more? `

- How do we put this into vscode's ssh-config? 
- Should this be on host or client?

Tried to add it, but without any luck.

**Comment by ColCh** [comment]
OK looks like I resolved this issue for myself. I want to tell you how I did it, so you could learn on my experience.

Of course, this is not a solution for this issue and cannot be used to close this issue :) I think it should be integrated into vscode itself. That will fix this whole issue.

## My setup

I use [Azure Virtual Desktop](https://azure.microsoft.com/en-us/products/virtual-desktop). (Actually, this is DevBox https://azure.microsoft.com/en-us/products/dev-box) with Windows on local notebook **and** remote machine, and I have **only RDP** connection to remote machine, without direct access to it (is uses gateways).

## What I did

I created WSL2 machine on remove Windows OS, started SSHD server on it, port forwarded SSH port to local machine, and **now I have uninterruptible SSH to remove machine's WSL2**.

_(please don't ask me why I have Windows and can't just install Ubuntu :) )_

## HOW:

**TLDR**: use devtunnels ( https://aka.ms/devtunnels ), proxy your SSH port via it, use localhost to connect to remotely forwarded port in vscode, add Timeout to ssh config for that host.

### Long info:

devtunnels ( https://aka.ms/devtunnels ) are using not really usual SSH, but it's own implementation, dev-tunnels-ssh ( https://github.com/microsoft/dev-tunnels-ssh/ ), which states this in the README:

> Supports [reconnecting](https://github.com/microsoft/dev-tunnels-ssh/blob/main/ProtocolExtensions.md) a disconnected session without disrupting channel streams.

So, after your computer goes from sleep, there is a delay about 5 seconds for WIFI, and some extra 3 seconds for reconnecting.

This breaks SSH, since default timeout is not that high. We gonna pipe those numbers up, by using this SSH config entries 

```
  ServerAliveInterval 60
  ServerAliveCountMax 120
```

60 seconds is high, it it will be enough. Plus, I will show how to specify this config specifically for your vscode host.

### Step by step

* Start sshd server (I will skip this please, for ubuntu go here https://ubuntu.com/server/docs/service-openssh. For WSL2 go for ubuntu instruction)
* Install devtunnels on local and remote machines: https://aka.ms/devtunnels 
* Login into devtunnel on remote machine and local machine `devtunnel user login`. Use **same** account on both machines.
* Start devtunnel host on remote machine. Example for proxying ports 22 and 8080: `devtunnel host -p 22 -p 8080`
  * NOTE: if you ware using WSL2 on server side as me, launch devtunnel from Windows side
* Remember ID of devtunnel, it will be like this: `Ready to accept connections for tunnel: THEREISNOSPOON` (THEREISNOSPOON is example ID)
* Connect to devtunnel on local machine: `devtunnel connect THEREISNOSPOON`
* Now to to VScode, click on `Add New SSH Host...`
* Enter your connection string, for my case it will be `ssh -C remoteuser@127.0.0.1 -p 22` (-p here because I'm using not 22 but 2222 port actually, but this makes no sense in this context)
* After it connected, open your ssh config file ( `~/.ssh/config` on ubuntu and `%HomePath%/.ssh/config`  on Windows )
* Optional, rename 127.0.0.1 config name. Add timeout entries to that host.


From:
```
Host 127.0.0.1
  HostName 127.0.0.1
  Compression yes
  User remoteuser
  Port 22
```

To: 
```
Host DEVBOX
  HostName 127.0.0.1
  Compression yes
  User remoteuser
  Port 22
  ServerAliveInterval 60
  ServerAliveCountMax 120
```

------

P.S. Sorry for too much info, I wanted it to be easily searchable on the internet.

-------

CC: @davidanthoff as topic starter

RELATED: #1790

**Comment by joshuataylor** [comment]
Great write up above, but you'll also want to tweak ClientAliveCountMax/ClientAliveInterval on the server, as the defaults don't last that long.

As you mentioned Ubuntu, see their [manpage for the specific version of openssh](https://manpages.ubuntu.com/manpages/jammy/man5/sshd_config.5.html).

tldr: change ClientAliveCountMax/ClientAliveInterval on the server to be a little higher than your client.

**Comment by starball5** [comment]
Related on Stack Overflow: [UNIX screen like behavior for Visual Studio Remote Development](https://stackoverflow.com/q/77721794/11107541)

**Comment by larsrgit** [comment]
For the time the code server keeps running, either the default time-out or when keeping it running forever using this solution https://github.com/microsoft/vscode-remote-release/issues/3096#issuecomment-1617150814, how can I reconnect to it? 

For me even if on the host the code server and all related processes are still running, when I reconnect to the host (even if the ssh connection is only interrupted for a few seconds), it starts a second code server on the host instead of reconnecting to the old one.

**Comment by pkubik** [comment]
Has anyone tried a simpler way to create a tunnel with https://code.visualstudio.com/docs/remote/tunnels ?
I used it on Remote window, but it only gave me access to my local workstation. Maybe it's worth a shot to install VSCode CLI on remote machine and use it there?

**Comment by abdelDarwish510** [comment]
Any updates on official support for this feature?

**Comment by vrtem** [comment]
hey, any updates here?

**Comment by dcnatsos** [comment]
Yeap, we need this feature!! And there must also be support for Jupyter Notebooks

**Comment by eliezedeck** [comment]
**AWS Cloud9** already has this persistent feature for years and years now. **Gitpod** also provides this via their own extension, however, it's tied to Gitpod servers (arguably cheaper than Azure Dev Box) and you need to configure timeouts to be very higher than default. So, **the possibility is there, it's just not getting enough attention**. I wonder why this stays stall for a long time, and yet it's got lots of demands.

**Comment by albertopasqualetto** [comment]
I think (but I still have not tested) that [here](https://github.com/microsoft/vscode/blob/35ac2d01a6b7ae2e0141377a6a924f050aa65237/cli/src/tunnels/code_server.rs#L489) is the problem

**Comment by eliezedeck** [comment]
> I think (but I still have not tested) that [here](https://github.com/microsoft/vscode/blob/35ac2d01a6b7ae2e0141377a6a924f050aa65237/cli/src/tunnels/code_server.rs#L489) is the problem

Interesting finding. So you're saying that you've tested it and it became persistent afterwards?

**Comment by marios1861** [comment]
Any updates on this? For remote AI training on GPU machines this is going to be such a huge quality of life improvement.

**Comment by youngdjn** [comment]
Another vote for this!

**Comment by wayhoww** [comment]
For those who consider persistent ssh session as a critically important feature, you can also try JetBrains Fleet. By utilizing its collaboration feature, it keeps terminals and debugging sessions after client being closed. Though it's not open source, and it does not even support plugins.

**Comment by moyamejiasr** [comment]
I also think we need this. And the worst part is how it is so easy to implement by making the stupid --enable-remote-auto-shutdown optional from the settings yet it has been 4 years and no progress.

**Comment by ionagamed** [comment]
Is there a confirmation from the team that PRs with a flag for `--enable-remote-auto-shutdown` would be accepted?

**Comment by marios1861** [comment]
@ionagamed The microsoft remote development extensions are not open source (see #30). From my exploration, the struct CodeServerArgs in [code_server.rs](https://github.com/microsoft/vscode/blob/a1c062b8622ef5da4fdd786030849439b233f043/cli/src/tunnels/code_server.rs#L50) needs to have an additional Option<bool> argument (e.g. keep_alive), which needs to be used in the [listen_on_port](https://github.com/microsoft/vscode/blob/a1c062b8622ef5da4fdd786030849439b233f043/cli/src/tunnels/code_server.rs#L486) function. To actually wire this back up to the extension settings, we would need permission to access and directly contribute to the actual extension. It's so infuriating that they have close-sourced the extension and simultaneously don't bother with SIMPLE quality of life improvements because they are focused on the (profitable) dev containers project, or the Microsoft hosted remote - tunnels, for which the rug can be pulled at any time. We should just write our own remote ssh dev extension.

**Comment by TomRaz** [comment]
That would be a great addition

**Comment by mrakgr** [comment]
I want this too.

**Comment by benz0id** [comment]
this would be great!

**Comment by PRO-2684** [comment]
Any news regarding this feature? This would be a great improvement. In my usecase, when I build a devcontainer, it might take *hours*, and I don't want to keep VSCode open on my laptop for that long - I'd like to close VSCode, do something else on my laptop, and reattach peoredically to check the progress. For now, as a *workaround*, I'm inspecting logs of devcontainer and running them manually with `nohup`, but this approach has certain minor short-comings.

**Comment by antonmx** [comment]
> One workround for now (only for users who wants the remote-server persistent FOREVER, not for users who wish to change the timeout value):
> 
> vscode starts the remote server by launching `sh <vscode-server-bin-path>/bin/code-server --start-server --host=127.0.0.1 --accept-server-license-terms --enable-remote-auto-shutdown --port=0 --telemetry-level all --connection-token-file <path-to-token-file>` in your default shell (`/bin/bash` in my case) after it established the SSH connection.
> 
> Therefore, just remove the `--enable-remote-auto-shutdown` argument, and the server will run without the idle timeout. This can be achieved by a customised wrapper for `sh`. My version goes like:
> 
>      #! /bin/bash
> 
>      args="$@"
> 
>      ignore_me='--enable-remote-auto-shutdown'
> 
>      if echo "$@" | grep -q -- "/.vscode-server"; then
>          if echo "$@" | grep -q -- "$ignore_me"; then
>              args=$(echo "$@" | sed "s/$ignore_me//g")
>          fi
>      fi
> 
>      /usr/bin/sh $args
> 
> And then `export PATH=<path-where-you-put-that-wrapper>:$PATH` in `.bashrc` (or into vscode's ssh-config, if you dislike polluting your shell environment). Now vscode will launch the remote-server without enabling the annoying idle-shutdown.


Thanks much for this idea! It did not work for me directly - I guess  .../.vscode-server is executed with "#!/usr/bin/env sh" which still points to /usr/bin/sh, rather than one in the PATH. I got it working via another dirty hack:

`sudo mv /usr/bin/sh /usr/bin/sh.real`

and creating slightly modified version of your suggestion in `/usr/bin/sh`:

    #!/bin/bash
 
    args="$@"
 
    ignore_me='--enable-remote-auto-shutdown'
 
    if echo "$@" | grep -q -- "/.vscode-server"; then
        if echo "$@" | grep -q -- "$ignore_me"; then
            args=$(echo "$@" | sed "s/$ignore_me//g")
        fi
    fi
 
    /usr/bin/sh.real $args

I have observed side effects as some errors reported on vscode startup, but no visible issues otherwise.

BTW, instead of using those two `if`'s in the script you can replace them with the single line:
`args=$( sed 's:\(.*/\.vscode-server.*\)$ignore_me\(.*\):\1\2:g'  <<< "$@" )`

I suspect in the above line better to quote `"$(...)"` as well as `"$args"` in the last line. Also in the last line I probably should use `eval`?

**Comment by Arcadiyyyyyyyy** [comment]
Please please please please please

**Comment by MTDickens** [comment]
Any updates on this long-lasting issue?

**Comment by edechamps-Google** [comment]
As part of my work I often find myself starting a long-running command (e.g. a build) in a VS Code Terminal, and then close my laptop and go to a meeting. When I open my laptop again 30 minutes later, I have to run the command again because the VS Code server bailed while I was away. This is quite annoying - at the very least the grace period should be customizable (I would personally set it to something like 3 hours).

**Comment by noreentry** [comment]
wow, this hardcoded constant really dictate my sleeping time - the 3h timeout is here:

https://github.com/microsoft/vscode/blame/71b706fe9f2561a0d4e1f1830b731b428e7962e6/src/vs/base/parts/ipc/common/ipc.net.ts#L298

monkey patch 3h->300h: on the server each time VSCode updates, in .vscode-server dir:

`grep -rl "ReconnectionGraceTime" . | xargs sed -i 's/ReconnectionGraceTime=108e5/ReconnectionGraceTime=108e7/g'`

**Comment by eevmanu** [comment]
in case anyone wants to visualize `grep` output better since the match contains lines so long, can run:

<details>
<summary> <code> $ grep -rPo ".{0,20}ReconnectionGraceTime=.{0,20}" ~/.vscode-server/ </code> </summary>

example

```sh
.vscode-server/cli/servers/Stable-7ad...925/server/out/vs/workbench/api/node/extensionHostProcess.js:]="TimeoutTime",e[e.ReconnectionGraceTime=108e5]="Reconnectio
.vscode-server/cli/servers/Stable-7ad...925/server/out/server-main.js:]="TimeoutTime",e[e.ReconnectionGraceTime=108e5]="Reconnectio
.vscode-server/cli/servers/Stable-488...f81/server/out/vs/workbench/api/node/extensionHostProcess.js:]="TimeoutTime",e[e.ReconnectionGraceTime=108e5]="Reconnectio
.vscode-server/cli/servers/Stable-488...f81/server/out/server-main.js:]="TimeoutTime",e[e.ReconnectionGraceTime=108e5]="Reconnectio
```

</details>

<details>
<summary> ℹ️  explanation of <code>grep</code> cli flags </summary>

`-P`: It enables Perl-Compatible Regular Expressions (PCRE), which gives us more powerful pattern matching capabilities, including the {min,max} quantifier.

`-o`: This tells grep to print only the matching parts of a line, rather than the full line. This is the magic that extracts our "snippet".

`".{0,20}ReconnectionGraceTime=.{0,20}"`: This is the PCRE pattern.
- `ReconnectionGraceTime`: The literal string you're searching for.
- `.`: Matches any single character (except newline).
- `{0,20}`: A quantifier meaning "match the preceding token (the .) between 0 and 20 times". We use {0,20} instead of {20} to gracefully handle cases where the match is at the very beginning or end of a line and doesn't have 20 characters of context available.

</details>

added the `=` at the end of `ReconnectionGraceTime` because I had some extra matches related to `ReconnectionGraceTimeIfNecessary()` which my guess is that they are not related

---

update: awesome command from @noreentry already manage the `=` character in the second part, in the `sed` command

**Comment by matifali** [comment]
## When the remote server is unavailable for multiple hours

I want to request the ability to auto-reconnect after the remote server is unavailable for an extended period (multiple hours). At https://github.com/coder/coder, a self-hosted CDE platform, users can schedule remote hosts to auto-start and auto-stop at specific times, reducing costs. A brief lifecycle for the remote workspace is described below.

### Remote Workspace lifecycle
- *Auto-stop*: Remote server shut down during off-hours (midnight)
- *Auto-start*: remote server resumes when needed (7 am)
- *Developer expectation*: Return to work seamlessly without losing state

### Current Pain Points:
- Manual "Reload Window" required every morning
- Unsaved changes lost during forced reload

**Comment by ibetitsmike** [comment]
@noreentry I'm not sure how that fix works for you. AFAICT vscode client actually hardcodes the 360 in https://github.com/microsoft/vscode/blob/4afd3e13ba536aba4d80f4ecf60b6be1332d3dc1/src/vs/platform/remote/common/remoteAgentConnection.ts#L678

I've opened a PR to allow for graceful reconnects on the vscode side: https://github.com/microsoft/vscode/pull/273060

**Comment by ruant** [comment]
Hopefully this can be solved soon, it's the most voted issue. 🤞

**Comment by tsuna** [comment]
I run remote-ssh over [quicssh](https://github.com/tsuna/quicssh) to work around this issue and also have my SSH sessions survive IP address changes that can happen along the way.

**Comment by fehmi** [comment]
Me also tired of manual reloading.

**Comment by CherryDT** [comment]
It is very annoying that this happens on WSL (not SSH - but issue #10767 was redirected here) after waking computer now all the time. It used to work fine.

**Comment by fehmi** [comment]
> Hopefully this can be solved soon, it's the most voted issue. 🤞

The issue is not even in the backlog yet.

**Comment by CherryDT** [comment]
For WSL it's a regression though... :(

**Comment by charles-cai** [comment]
lost in the chat history, so 6 years down the line, where are we? 

probably devtunnels is the way to go? https://aka.ms/devtunnels above which was posted in 2023, sigh

**Comment by youngdjn** [comment]
For data scientists using Python or R specifically, the excellent VS Code fork [Positron](https://positron.posit.co/) supports [long-running SSH sessions](https://positron.posit.co/remote-ssh.html#long-running-sessions).

**Comment by joelsandocu** [comment]
This would be a very useful feature, +1,000,000 votes!

**Comment by tsuna** [comment]
I could implement this easily if the extension was open source. Does anyone know why it's kept closed source?

**Comment by cristianocca** [comment]
+100
